### DISCUSSION TITLE
Game Subdirectory Layout

### DISCUSSION THREAD

**Discussion by randyhook** [evennia/evennia#3031] [Evennia Questions] [answered]
In the documentation, under Beginner Tutorial | Part 3 | Code Structure and Utilities,  it is suggested that within your game directory (in this case mygame/), you create a subdirectory for your game modules (in this case edadventure/). In this directory you house your game-specific modules and tests.

I'm trying to figure out why you would want this subdirectory and not just have mygame/tests, mygame/typeclasses/myobjects/ and so forth. Is it just a matter of organizational preference or is there an underlying reason for this hierarchy? At first glance it's almost like the reason would be to house multiple games within the mygame directory.

**Comment by Aesran** [comment]
Not sure, but i think it's because when you're going to update evennia, it's easier to have your modification away from those of the game.
You don"t make update in directory from the github, you have your.

It's make updating more easy and to be honest, more clear.

**Reply by InspectorCaracal** [reply]
Just for a bit of clarity: the `mygame` directory mentioned in the original post is already a separate directory from evennia's own code. Evennia is designed so that you make all of your own code changes without ever touching the core engine's files.

**Comment by randyhook** [comment]
I figured it might be something like that, but from what I'm seeing, you can make modifications to the code anywhere in your mygame/ directory. In fact, that same directory is also where you create your own web portal files, and that is in mygame/web/website, so it doesn't seem like the mygame/ directory is "safe" if it were to be affected by an update.

**Comment by InspectorCaracal** [comment]
The EvAdventure tutorial is still a WIP, according to Griatch (and to the number of typos in it), so it's possible that's also something that'll get revised and it'll suggest you make a gamedir of `evadventure`. If not, my guess is it's so that you can build it as a separate internal module to practice on and reference from without affecting the actual game.

Full disclaimer, though, I haven't actually gone through the new tutorial pages, so I haven't seen this part yet.

**Comment by randyhook** [comment]
OK, I'll go with that. Thanks to both for the info.

**Comment by Griatch** [comment, marked as answer]
@randyhook 

@InspectorCaracal is very close to the official answer :) The tutorial is very much WIP (see big red warning and notes all over).

For your own game, you don't need to make a separate folder with code like this - you can indeed just modify the 'templates' to do things directly. Doing so doesn't affect your updating procedure - the game dir is separate from Evennia, it's up to you to customize to your liking. I don't actually mention this in the folder-layout lesson, I'll do so as I revise this. 

There is however a practical reason for me choosing to make a  `mygame/evadventure` folder rather than embedding this code in the templates we provide by default: The reason is that as I develop the tutorial I am simultaneously creating the `evennia/contrib/tutorials/evadventure/` folder as a functioning implementation of said tutorial. People learn in different ways. The idea is that (once the tutorial is finished), you as a reader can use the tutorial in different ways: 

- Read the tutorial and implement each part as written, using the ready-made code to check your progress
- Treat the tutorial code as a collection of contribs - copy & paste just the code you need, using the tutorial text just as extra documentation
- Copy the whole `contrib/tutorials/evadventure` folder into your game dir and use it as the base for your game or as a learning project for your own studies, using the tutorial text only as a reference.

So the tutorial and contrib code hangs together, and it makes sense to group it all in one folder. It also allows for another useful lesson: Towards the end of part 3 (not yet written), we will tie everything together by hooking all those custom typeclasses and commands into Evennia. This means adding the relevant path to settings - another useful skill to have. 

That's my thinking at the moment at least. If it turns out that making the tutorial code in a separate folder is just confusing the issue, I can rethink. At any rate, I should make a sidebar in the tutorial to clarify that this is not required for your own game (you can really organize your game dir however you like).

**Comment by randyhook** [comment]
Excellent! Thanks so much for the clarification.
